### PR TITLE
add topology and bootstrap

### PR DESCRIPTION
**Pull request by lxning**
This diff includes the following parts:

1. dir cluster:
- topology 
- common api related to information from etcd 

2. dir datanode:
- bootstrap manager
- datanode

3. Todo comments in cmd/aresd/cmd/cmd.go for later bootstrap integration

### CODE REVIEW

File: datanode/topology/dynamic.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	i.topo = topo
+	return i.topo, nil
+}
+
+func (i *dynamicInitializer) TopologyIsSet() (bool, error) {
+	return true, nil
```
Comment (shz117) [inline datanode/topology/dynamic.go]: nit this function is dummy.. should this be like `return i.topo != nil, nil` at least?

File: datanode/bootstrap/types.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
+package bootstrap
+
+import (
+	"github.com/uber/aresdb/datanode/topology"
+	"time"
+	"github.com/m3db/m3/src/dbnode/storage/bootstrap/result"
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/bootstrap/types.go]: this diff is only for topology part, not bootstrap part is only for compile

File: datanode/bootstrap/types.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	// an error should it want to entirely cancel the bootstrapping of the node,
+	// i.e. non-recoverable situation like not being able to read from the filesystem.
+	ReadData(
+		ns namespace.Metadata,
+		shardsTimeRanges result.ShardTimeRanges,
+		runOpts RunOptions,
+	) (result.DataBootstrapResult, error)
+
+	// AvailableIndex returns what time ranges are available for bootstrapping.
+	AvailableIndex(
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/bootstrap/types.go]: same

File: datanode/bootstrap/types.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+// with the mindset that it will always be set to default values from the constructor.
+type Process interface {
+	// Run runs the bootstrap process, returning the bootstrap result and any error encountered.
+	Run(start time.Time, ns namespace.Metadata, shards []int) (ProcessResult, error)
+}
+
+// ProcessResult is the result of a bootstrap process.
+type ProcessResult struct {
+	DataResult  result.DataBootstrapResult
+	IndexResult result.IndexBootstrapResult
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/bootstrap/types.go]: same

File: datanode/types.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+// DataNode includes metastore, memstore and diskstore
+type DataNode struct {
+	metastore.MetaStore
+	memstore.MemStore
+	diskstore.DiskStore
+}
+
+// databaseMediator mediates actions among various datanode managers.
+type datanodeMediator interface {
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/types.go]: same

File: utils/options.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/uber-go/tally"
+	"github.com/uber/aresdb/common"
+)
+
+const (
+	defaultSamplingRate = 1.0
+)
+
+// Options represents the options for instrumentation.
+type Options interface {
```
[the remark this answers is not in the document]
Comment (lxning) [reply utils/options.go]: it is used for chaining configurations to make code clean.

File: datanode/topology/types.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+	ServicesOverrideOptions() services.OverrideOptions
+
+	// SetQueryOptions sets the ConfigService query options
+	SetQueryOptions(value services.QueryOptions) DynamicOptions
+
+	// QueryOptions returns the ConfigService query options
+	QueryOptions() services.QueryOptions
+
+	// SetInstrumentOptions sets the instrumentation options
+	SetInstrumentOptions(value utils.Options) DynamicOptions
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/topology/types.go]: same

File: datanode/bootstrap/types.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	// Provide constructs a bootstrap process.
+	Provide() (Process, error)
+}
+
+// Process represents the bootstrap process. Note that a bootstrap process can and will
+// be reused so it is important to not rely on state stored in the bootstrap itself
+// with the mindset that it will always be set to default values from the constructor.
+type Process interface {
+	// Run runs the bootstrap process, returning the bootstrap result and any error encountered.
+	Run(start time.Time, ns namespace.Metadata, shards []int) (ProcessResult, error)
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/bootstrap/types.go]: same

File: datanode/options.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+var (
+	// defaultBootstrapProcessProvider is the default bootstrap provider for the database.
+	// TODO: defaultBootstrapProcessProvider = bootstrap.NewNoOpProcessProvider()
+	timeZero                        time.Time
+)
+
+var (
+	errNamespaceInitializerNotSet = errors.New("namespace registry initializer not set")
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/options.go]: this error should belong to namespace registration process.

File: datanode/options.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	"errors"
+	"github.com/uber/aresdb/datanode/bootstrap"
+	"github.com/uber/aresdb/utils"
+	"time"
+)
+
+var (
+	// defaultBootstrapProcessProvider is the default bootstrap provider for the database.
+	// TODO: defaultBootstrapProcessProvider = bootstrap.NewNoOpProcessProvider()
+	timeZero                        time.Time
```
[the remark this answers is not in the document]
Comment (lxning) [reply datanode/options.go]: removed

File: cluster/types.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
+package cluster
+
+import (
+	"github.com/m3db/m3/src/dbnode/namespace"
```
Comment (jshencode) [inline cluster/types.go]: let's not depend on  m3's dbnode package

**Pull request closed without merging**